File: Data Structures & Algorithms/palindrome-partitioning/submission-0.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        res = []

        def isPalindrome(curr):
            l, r = 0, len(curr) - 1

            while l < r:
                if curr[r] != curr[l]:
                    return False
                
                r -= 1
                l += 1
            
            return True
        
        def backtrack(curr, i):
            if i == len(s):
                res.append(curr.copy())
                return

            #Check if substring is palindrome
            j = i + 1
            while j <= len(s):
                if isPalindrome(s[i:j]):
                    curr.append(s[i:j])
                    backtrack(curr, j)
                    curr.pop()
                j += 1

        backtrack([], 0)
        return res

            
        

File: Data Structures & Algorithms/rotting-fruit/submission-0.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int res = 0;
        int fresh = 0;
        queue<pair<int, int>> q;


        for(int row = 0; row < grid.size(); row++){
            for(int col = 0; col < grid[0].size(); col++){
                if(grid[row][col] == 1){
                    fresh++;
                }
                if(grid[row][col] == 2){
                    q.push({row, col});
                }
            }
        }

        vector<pair<int, int>> directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        while(!q.empty() && fresh > 0){
            int qlength = q.size();

            for(int i = 0; i < qlength; i++){
                auto cell = q.front();
                q.pop();

                for(const auto& curr : directions){
                    int row = cell.first + curr.first;
                    int col = cell.second + curr.second;

                    //Check if in bounds and orange
                    if(row >= 0 && row < grid.size() && col >= 0 && col < grid[0].size() && grid[row][col] == 1){
                        grid[row][col] = 2;
                        q.push({row, col});
                        fresh--;
                    }
                }
            }
            res++;
        }
        if(fresh > 0){
            return -1;
        }
        return res;
    }
};
